### PR TITLE
Hotfix/syal 1137

### PR DESCRIPTION
**Pull request by HieuLM-AVISPL**

### CODE REVIEW

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java [hunk 1]
```diff
@@ -1352,16 +1383,107 @@ void populateRegionDetails(AggregatedDevice aggregatedDevice) throws Exception {
...
-			populateOccupancyData(properties, controls, currentDate, yesterdayDate, deviceId, buildingID, floorName, deviceName);
+			CompletableFuture<Boolean> occupancyFuture = CompletableFuture.supplyAsync(
+					() -> populateOccupancyData(properties, controls, currentDate, yesterdayDate, deviceId, buildingID, floorName, deviceName));
 			// Retrieve region tags
-			populateRegionTag(properties, controls, deviceId);
+			CompletableFuture<Boolean> regionTagFuture = CompletableFuture.supplyAsync(
+					() -> populateRegionTag(properties, controls, deviceId));
+
+			// retry on 429 error
+			occupancyFuture.thenApply(result -> {
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:1438]: haven't you considered using an executor service with a limited thread pool for this? It'd be easier to control these + terminate operations on adapter destroy action

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java [hunk 2]
```diff
@@ -1699,21 +1850,24 @@ private void populateOccupancyData(Map<String, String> properties, List<Advanced
...
 					if (occupancyRegionResponses.length == 0) {
 						properties.put(SembientAggregatorConstant.PROPERTY_MESSAGE, SembientAggregatorConstant.NO_DATA);
-						return;
 					}
 					dateToBeDisplayed = yesterdayDate;
 				} else {
 					properties.put(SembientAggregatorConstant.PROPERTY_MESSAGE, SembientAggregatorConstant.NO_DATA);
-					return;
+					if (cachedTooManyRequestError.contains(secondRequest)) {
+						cachedTooManyRequestError.remove(secondRequest);
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: would make sense to clear this list on destroy event, not to wait until gc gets to it :)

Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: maybe it would make sense to clear this list on destroy event, not to wait until gc gets to it :)

Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: maybe it would make sense to clear this list on destroy event, not to wait until gc gets to it :)

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java [hunk 3]
```diff
@@ -1774,22 +1927,26 @@ private void populateOccupancyData(Map<String, String> properties, List<Advanced
 			if (!properties.containsKey(SembientAggregatorConstant.PROPERTY_HOUR)) {
 				properties.put(SembientAggregatorConstant.PROPERTY_MESSAGE, SembientAggregatorConstant.NO_DATA);
 			}
+			if (cachedTooManyRequestError.contains(firstRequest)) {
+				cachedTooManyRequestError.remove(firstRequest);
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: there's no need to check for "contains" since remove won't remove anything if there's no match :)

**Review by MaksimR-AVISPL — CHANGES_REQUESTED** [review]

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java (hunk 3 above)
Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java (hunk 1 above)
Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:1438]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java (hunk 2 above)
Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: I fixed it

Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: I fixed it

Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java [hunk 4]
```diff
@@ -1109,28 +1144,28 @@ private void filterByFloors(String buildingID, String buildingName, String[] flo
...
 						// Filter by region type
-						if (StringUtils.isNotNullOrEmpty(deviceTypeFilter)) {
-							String[] listTypeToBeFilter = deviceTypeFilter.split(SembientAggregatorConstant.COMMA);
-							for (String type : listTypeToBeFilter) {
-								retrieveDevices(buildingID, buildingName, floor, type.trim());
-							}
-						} else {
-							retrieveDevices(buildingID, buildingName, floor, null);
-						}
+//						if (StringUtils.isNotNullOrEmpty(deviceTypeFilter)) {
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: Please remove the commented-out segments. If we need them back - we can always revert the changes :)

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java [hunk 5]
```diff
@@ -1109,28 +1144,28 @@ private void filterByFloors(String buildingID, String buildingName, String[] flo
...
 			for (String floorName : floorNames) {
-				if (StringUtils.isNotNullOrEmpty(deviceTypeFilter)) {
-					String[] listTypeToBeFilter = deviceTypeFilter.split(SembientAggregatorConstant.COMMA);
-					for (String type : listTypeToBeFilter) {
-						retrieveDevices(buildingID, buildingName, floorName, type.trim());
-					}
-				} else {
-					retrieveDevices(buildingID, buildingName, floorName, null);
-				}
+//				if (StringUtils.isNotNullOrEmpty(deviceTypeFilter)) {
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: This section also

**Review by MaksimR-AVISPL — APPROVED** [review]

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java (hunk 4 above)
Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java (hunk 5 above)
Comment (HieuLM-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java]: I fixed it
